tempesta_fw/t/functional/regression/test_health_monitor.py

```diff
@@ -0,0 +1,227 @@
+                self.client, ch_enabled,
+                True
+            ),
+            Stage(
+                self.tester, self.tempesta,
+                self.client, ch_disabled,
+                False
+            ),
+            Stage(
+                self.tester, self.tempesta,
+                self.client, ch_enabled,
+                True, trans=True
+            ),
+            Stage(
+                self.tester, self.tempesta,
+                self.client, ch_disabled,
+                False
+            ),
+            Stage(
+                self.tester, self.tempesta,
+                self.client, ch_enabled,
+                True, trans=True
+            )
+        ])
+
+    def assert_tempesta(self):
+        self.tester.assert_stages()
+        err_msg = 'Tempesta must have errors due' \
+                  ' to inability to schedule request'
+        functional.FunctionalTest.assert_tempesta(self)
+        self.assertTrue(self.tempesta.stats.cl_msg_other_errors > 0,
+                        msg=err_msg)
+
+    def test(self):
+        config = (
+            'server_failover_http 404 300 10;\n'
+            'health_check h_monitor1 {\n'
+            'request "GET / HTTP/1.1\r\nHost: example.com\r\n\r\n";\n'
+            'request_url "/page.html";\n'
+            'resp_code 200;\n'
+            'resp_crc32 3456;\n'
+            'timeout 15;\n'
+            '}\n'
+            'cache 0;\n'
+        )
+        self.generic_test_routine(config, self.messages)
+
+
+CHAIN_TIMEOUT = 100
+
+class StagedDeproxy(deproxy.Deproxy):
+
+    def __init__(self, *args, **kwargs):
+        deproxy.Deproxy.__init__(self, *args, **kwargs)
+        self.stages = []
+        self.stages_n = 0
+        self.stages_processed = 0
+        self.current_stage = None
+
+    def run(self):
+        for _ in range(self.message_chains):
+            self.prepare()
+            self.loop()
+            if not self.check_expectations():
+                break
+
+    def prepare(self):
+        self.current_stage.prepare()
+
+    def check_expectations(self):
+        return self.current_stage.check_results()
+
+    def assert_stages(self):
+        assert self.stages_processed == self.stages_n, \
+            ("Not all stages are passed: processed"
+             " stages = %d, total stages count = %d"
+             % (self.stages_processed, self.stages_n))
+
+    def set_stages(self, stages):
+        self.stages = stages
+        self.stages_n = len(stages)
+        self.current_stage = self.stages.pop(0)
+        self.stages_processed = 1
+
+    def next_stage(self, messages):
+        if not self.stages:
+            return False
+        self.stages[0].check_transition(messages)
+        self.current_stage = self.stages.pop(0)
+        self.stages_processed += 1
+        return True
+
+    def get_server_path(self):
+        return 'default/%s:%s' % (self.servers[0].ip, self.servers[0].port)
+
+
+class Stage(object):
```

Comment:
I know this isn't C, but please keep forward declaration. Having most experience in C it's much easier to read code which firstly defines all the staff and after that uses it.

Reply:
Corrected.